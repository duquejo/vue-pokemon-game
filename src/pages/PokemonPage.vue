<template>
  <h1 v-if="!pokemon">Wait a sec...</h1>
  <div v-else>
    <h1>Who's that Pok√©mon?</h1>
    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"/>
    <PokemonOptions
      :options="pokemonArray"
      @selection="checkAnswer"/>

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button class="fade-in" @click="newGame">New game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';
import getPokemonOptions from '../helpers/getPokemonOptions';

export default {
  name: 'PokemonPage',
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArray: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArray = await getPokemonOptions();
      this.pokemon = this.pokemonArray[Math.floor(Math.random() * 4)];
    },
    checkAnswer(pokeId) {
      this.showPokemon = true;
      this.showAnswer = true;
      if( pokeId === this.pokemon.id ) {
        this.message = `That\'s correct!, it's a ${this.pokemon.name}!`;
      } else {
        this.message = `Oops, it's a ${this.pokemon.name}!`;
      }
    },
    newGame() {
      this.showAnswer = false;
      this.showPokemon = false;
      this.pokemon = null;
      this.message = '';
      this.pokemonArray = [];
      this.mixPokemonArray();
    }
  },
  mounted() {
    this.mixPokemonArray(); // Retrieve pokemons when the component is mounted
  }
}
</script>

<style scoped>
button {
  background-color: #FFF;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  padding: 10px;
}
button:hover {
  background-color: rgba(0, 0, 0, 0.05);
}
</style>