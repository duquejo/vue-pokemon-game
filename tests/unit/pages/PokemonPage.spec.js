import PokemonPage from '@/pages/PokemonPage';
import { mount, shallowMount } from '@vue/test-utils';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage.vue', () => {

  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  it('should match the snapshot.', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call the mixPokemonArray after mount', () => {
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray'); // Must have to instanciate this spy first.
    shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  it('should match the snapshot when the PokÃ©mons are loaded', () => {
    /**
     * Mounts every inner component dependencies
     */
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show the PokemonPicture and PokemonOptions components', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        }
      }
    });

    const pictureContainer = wrapper.find('pokemon-picture-stub');
    const optionsContainer = wrapper.find('pokemon-options-stub');

    expect(pictureContainer.exists()).toBeTruthy();
    expect(optionsContainer.exists()).toBeTruthy();

    expect(pictureContainer.attributes('pokemonid')).toBe(String(pokemons[0].id));
    expect(optionsContainer.attributes('options')).toBeTruthy();
  });


  it('checkAnswer - Should change the state.', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArray: pokemons,
          pokemon: pokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: '',
        }
      }
    });

    await wrapper.vm.checkAnswer(pokemons[1].id);

    /**
     * From the DOM
     */
    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.find('h2').text()).toBe(`That\'s correct!, it\'s a ${pokemons[1].name}!`);
    expect(wrapper.vm.showPokemon).toBeTruthy();

    await wrapper.vm.checkAnswer(10);

    /**
     * From the data
     */
    expect(wrapper.vm.message).toBe(`Oops, it\'s a ${pokemons[1].name}!`);
  });
});