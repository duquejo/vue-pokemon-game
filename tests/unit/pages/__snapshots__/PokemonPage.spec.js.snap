// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage.vue should match the snapshot when the Pokémons are loaded 1`] = `
<div>
  <h1>Who's that Pokémon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="Pokémon"></div>
  <div class="options-container">
    <ul>
      <li>pikachu</li>
      <li>charmander</li>
      <li>ivysaur</li>
      <li>voltorb</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;

exports[`PokemonPage.vue should match the snapshot. 1`] = `<h1>Wait a sec...</h1>`;
